st3/lsp_utils/npm_client_handler.py

```diff
@@ -68,22 +67,14 @@ def cleanup(cls) -> None:
         if cls.__server:
             cls.__server.cleanup()
 
-    @property
-    def name(self) -> str:
-        return self.package_name.lower()  # type: ignore
-
-    @property
-    def config(self) -> ClientConfig:
-        assert self.__server
-
-        configuration = {
-            'enabled': True,
-            'command': ['node', self.__server.binary_path] + self.get_binary_arguments(),
-        }
+    @classmethod
+    def name(cls) -> str:
+        cls.setup()
```

Comment:
The call for setup doesn't feel right here. 

Is there a reason why you called setup in a method that should return a name 🙂

Reply:
Because: https://github.com/sublimelsp/LSP/issues/899

Reply:
Why don't we define a `setup` classmethod on the Abstract class?
```
class AbstractPlugin(metaclass=ABCMeta):
    @classmethod
    @abstractmethod
    def setup(cls) -> None:
        """
        Do the setup
        """
        # not mandatory to implement
        pass 
```
than in register_plugin before returning a Client config call plugin.setup()
```diff

def register_plugin(plugin: Type[AbstractPlugin], update_global_configs: bool = True) -> None:
    global _plugins
    global client_configs

+   plugin.setup()
    try:
        config = ClientConfig(
            name=plugin.name(),
```

Would that work?

Reply:
Since `setup` can be async and run on separate thread, it should ideally not be called during `register_plugin`. `register_plugin` should be called only after `setup` has been fully done.